File: netlify/functions/test_backend/ldr_to_lbcode.py
# ldr_to_lbcode.py
import sys
from math import isclose

LDU_WIDTH = 20
LDU_DEPTH = 20
LDU_HEIGHT = 24

# Rotation matrices as in LDRConv.cpp
TOWARDS_X       = ( 1.0, 0.0, 0.0,
                    0.0, 1.0, 0.0,
                    0.0, 0.0, 1.0)

TOWARDS_X_ALT   = (-1.0, 0.0, 0.0,
                    0.0, 1.0, 0.0,
                    0.0, 0.0,-1.0)

TOWARDS_Y       = ( 0.0, 0.0, 1.0,
                    0.0, 1.0, 0.0,
                   -1.0, 0.0, 0.0)

TOWARDS_Y_ALT   = ( 0.0, 0.0,-1.0,
                    0.0, 1.0, 0.0,
                    1.0, 0.0, 0.0)


def floats_equal_matrix(m, ref, eps=1e-6):
    """Compare two 3x3 matrices (9 floats) with a small tolerance."""
    return all(isclose(m[i], ref[i], abs_tol=eps) for i in range(9))


def parse_type1_line(line):
    """
    Parse a type-1 LDraw line.
    Returns (x, y, z, [9 matrix floats]) or None if the line is not valid / not type 1.
    Expected simplified format:
      1 color x y z a b c d e f g h i <filename>
    """
    parts = line.strip().split()
    if not parts:
        return None

    # Line type
    try:
        t = int(parts[0])
    except ValueError:
        return None

    if t != 1:
        # Ignore non-type-1 lines (same behavior as LDRConv)
        return None

    # We need at least: 1, color, x, y, z, then 9 matrix elements
    if len(parts) < 14:
        return None

    try:
        # We ignore the color for the conversion
        # color = int(parts[1])
        x = float(parts[2])
        y = float(parts[3])
        z = float(parts[4])

        # 3x3 matrix from parts[5] ... parts[13]
        m = [float(parts[5 + i]) for i in range(9)]
    except ValueError:
        return None

    return x, y, z, m


def convert(input_path, output_path):
    # Each entry: (rot, grid_x, grid_y, grid_z)
    parts = []

    # Bounds used to recenter the model
    min_x = float("inf")
    max_x = float("-inf")
    min_y = float("inf")
    max_y = float("-inf")

    with open(input_path, "r") as f:
        for line in f:
            parsed = parse_type1_line(line)
            if parsed is None:
                continue

            x, y, z, m = parsed

            # Determine orientation (same logic as LDRConv with is_towards_x)
            if floats_equal_matrix(m, TOWARDS_X) or floats_equal_matrix(m, TOWARDS_X_ALT):
                is_towards_x = True
            elif floats_equal_matrix(m, TOWARDS_Y) or floats_equal_matrix(m, TOWARDS_Y_ALT):
                is_towards_x = False
            else:
                # Unsupported orientation: skip this brick
                # (C++ version would return an error instead)
                continue

            # Coordinate mapping LDraw → printer space
            # Studio   LDraw   Printer
            #  +X       +X        +X
            #  +Y       +Y        -Z
            #  +Z       -Z        +Y
            pos_x = round(x)
            pos_y = round(-z)
            pos_z = round(-y)

            # Discrete grid conversion (same idea as ConvertLDRPosition)
            if is_towards_x:
                lx = int((pos_x - 2 * LDU_WIDTH) / LDU_WIDTH)
                ly = int((pos_y - LDU_DEPTH) / LDU_DEPTH)
            else:
                lx = int((pos_x - LDU_WIDTH) / LDU_WIDTH)
                ly = int((pos_y + LDU_DEPTH) / LDU_DEPTH)

            lz = int((pos_z - LDU_HEIGHT) / LDU_HEIGHT)

            rot = 0 if is_towards_x else 1

            parts.append((rot, lx, ly, lz))

            # Update area bounds (equivalent to UpdateContextArea)
            if is_towards_x:
                width = 4
                depth = 2
                part_min_y = ly
                part_max_y = ly + depth
            else:
                width = 2
                depth = 4
                # Note: dpos.y - (depth - 1) in C++
                part_min_y = ly - (depth - 1)
                part_max_y = ly

            part_min_x = lx
            part_max_x = lx + width

            min_x = min(min_x, part_min_x)
            max_x = max(max_x, part_max_x)
            min_y = min(min_y, part_min_y)
            max_y = max(max_y, part_max_y)

    if not parts:
        # No valid bricks in the file
        raise RuntimeError("No valid bricks found in LDR file.")

    # Recenter like in DoConversion (p.pos.x -= ctx.min_x, p.pos.y -= ctx.min_y)
    recentered = []
    for rot, lx, ly, lz in parts:
        recentered.append((rot, lx - int(min_x), ly - int(min_y), lz))

    # Sort by layer: z, then x, then y (same as PartInfo::operator< in LDRConv)
    # (rot, x, y, z) → sort key (z, x, y)
    recentered.sort(key=lambda p: (p[3], p[1], p[2]))

    # Final width/height after recentering
    width_final = int(max_x - min_x)
    height_final = int(max_y - min_y)

    # Write .lbcode file
    with open(output_path, "wb") as out:
        # Header: max_x, max_y (after recentering)
        out.write(width_final.to_bytes(2, byteorder="little", signed=False))
        out.write(height_final.to_bytes(2, byteorder="little", signed=False))

        # Then each brick: 1 byte rot, 2 bytes x, 2 bytes y, 2 bytes z
        for rot, x, y, z in recentered:
            out.write(bytes([rot]))
            out.write(int(x).to_bytes(2, byteorder="little", signed=True))
            out.write(int(y).to_bytes(2, byteorder="little", signed=True))
            out.write(int(z).to_bytes(2, byteorder="little", signed=True))


if __name__ == "__main__":
    if len(sys.argv) != 3:
        print("Usage: python ldr_to_lbcode.py <input.ldr> <output.lbcode>")
        sys.exit(1)
    convert(sys.argv[1], sys.argv[2])

File: netlify/functions/test_backend/ldr_to_lbcode.js

function arrayEquals(a, b) {
    if (a === b) return true; 
    if (a == null || b == null) return false;
    if (a.length !== b.length) return false;

    for (var i = 0; i < a.length; ++i) {
        if (a[i] !== b[i]) return false;
    }
    return true;
}

function CompareParts(a, b) {
    if (a.pos.z < b.pos.z) {
        return -1;
    }
    else if (a.pos.z == b.pos.z) {
        if (a.pos.x < b.pos.x) {
            return -1;
        }
        else if (a.pos.x == b.pos.x) {
            if (a.pos.y < b.pos.y) {
                return -1;
            }
            else if (a.pos.y == b.pos.y) {
                return 0;
            }
            else {
                return 1;
            }
        }
        else {
            return 1;
        }
    }
    else {
        return 1;
    }
}

function ConvertLDRPosition(apos, is_towards_x) {
    const LDU_width = 20;
    const LDU_depth = 20;
    const LDU_height = 24;

    return is_towards_x ? {
        x: (apos.x - 2 * LDU_width) / LDU_width,
        y: (apos.y - LDU_depth) / LDU_depth,
        z: (apos.z - LDU_height) / LDU_height
    } : {
        x: (apos.x - LDU_width) / LDU_width,
        y: (apos.y + LDU_depth) / LDU_depth,
        z: (apos.z - LDU_height) / LDU_height
    };
}

function UpdateContextArea(is_towards_x, dpos, ctx) {
    const X_WIDTH = 4; 
    const X_DEPTH = 2; 
    const Y_WIDTH = 2; 
    const Y_DEPTH = 4; 

    width = is_towards_x ? X_WIDTH : Y_WIDTH; 
    depth = is_towards_x ? X_DEPTH : Y_DEPTH; 

    min_x = dpos.x; 
    max_x = dpos.x + width; 
    min_y = dpos.y - ((!is_towards_x) ? depth - 1 : 0); 
    max_y = dpos.y + (is_towards_x ? depth : 0);

    ctx.min_x = Math.min(ctx.min_x, min_x);
    ctx.max_x = Math.max(ctx.max_x, max_x);
    ctx.min_y = Math.min(ctx.min_y, min_y);
    ctx.max_y = Math.max(ctx.max_y, max_y);

    return ctx;  
}

function ProcessPosition(p, ctx) {
    const towards_x = [ 1, 0, 0, 0, 1, 0, 0, 0, 1 ];
    const towards_x_alt = [ -1, 0, 0, 0, 1, 0, 0, 0, -1 ];
    const towards_y = [ 0, 0, 1, 0, 1, 0, -1, 0, 0 ];
    const towards_y_alt = [ 0, 0, -1, 0, 1, 0, 1, 0, 0 ];

    let is_towards_x = false;

    if (arrayEquals(p.rotation, towards_x) || arrayEquals(p.rotation, towards_x_alt)) {
        is_towards_x = true;
    }
    else if (arrayEquals(p.rotation, towards_y) || arrayEquals(p.rotation, towards_y_alt)) {
        is_towards_x = false;
    }
    else {
        throw new Error("Unsupported rotation");
    }
    
    let apos = {
        x: Math.round(p.x),
        y: Math.round(-p.z),
        z: Math.round(-p.y)
    };

    let dpos = ConvertLDRPosition(apos, is_towards_x);

    let encoded_rotation = is_towards_x ? 0 : 1;
    
    const DEFAULT_COLOR = 15;

    ctx.parts.push({
        rotation: encoded_rotation,
        pos: dpos,
        is_red: p.color == DEFAULT_COLOR
    });

    ctx = UpdateContextArea(is_towards_x, dpos, ctx);

    return ctx;
}

function ProcessLine(line, ctx) {
    const MATRIX_ELEMENTS_COUNT = 12;

    var type;
    var color;

    elements = line.split(" ");

    type = parseInt(elements[0]);

    if (type != 1) {
        if (type != 1) {
            console.log(`Warning (line ${ctx.line_no}): Unsupported type detect, ignoring line`);
            return ctx;
        }
    }
    
    if (elements.length != 15) {
        throw new Error(`Invalid line format (line ${ctx.line_no})`);
    }

    color = parseInt(elements[1]);

    matrix = [];

    for (var i = 0; i < MATRIX_ELEMENTS_COUNT; i++) {
        matrix.push(parseFloat(elements[i + 2]));
    }

    let p = {
        x: matrix[0],
        y: matrix[1],
        z: matrix[2],
        rotation: [
            matrix[4], matrix[5], matrix[6],
            matrix[8], matrix[9], matrix[10],
            matrix[12], matrix[13], matrix[14]
        ],
        color: color
    };

    return ProcessPosition(p, ctx);
}

function DoConversion(lines, ctx) {
    ctx.line_no = 1; 
    
    for (i = 0; i < lines.length; i++) {
        ctx = ProcessLine(line, ctx);
        ctx.line_no += 1;
    }

    ctx.parts.sort(CompareParts);

    let buffer = Buffer.from([]);

    for (var p of ctx.parts) {
        p.pos.x -= ctx.min_x;
        p.pos.y -= ctx.min_y;

        let encoded_rotation = p.rotation | (p.is_red ? 0 : 2);

        let pbuf = new Uint8Array(7);
        pbuf[0] = encoded_rotation;
        pbuf[1] = p.pos.x & 0xFF;
        pbuf[2] = (p.pos.x >> 8) & 0xFF;
        pbuf[3] = p.pos.y & 0xFF;
        pbuf[4] = (p.pos.y >> 8) & 0xFF;
        pbuf[5] = p.pos.z & 0xFF;
        pbuf[6] = (p.pos.z >> 8) & 0xFF;

        buffer = Buffer.concat([buffer, Buffer.from(pbuf)]);
    }

    ctx.max_x -= ctx.min_x;
    ctx.max_y -= ctx.min_y;
    ctx.raw_buffer = buffer;

    return ctx;
}

function ConvertLDR(lines) {
    let ctx = {
        raw_buffer: [],
        line_no: 0,
        min_x: Infinity,
        max_x: -Infinity,
        min_y: Infinity,
        max_y: -Infinity,
        parts: []
    };

    ctx = DoConversion(lines, ctx);

    let metadata = new Uint16Array(2);
    metadata[0] = ctx.max_x & 0xFFFF;
    metadata[1] = ctx.max_y & 0xFFFF;

    return Buffer.concat([Buffer.from(metadata.buffer), ctx.raw_buffer]);
}
